File: app/src/main/kotlin/ru/ktoddler/model/repository/LocalRepository.java
package ru.ktoddler.model.repository;

import android.support.annotation.NonNull;

import com.annimon.stream.Optional;

import java.util.List;

import io.reactivex.Observable;


public interface LocalRepository {

    @NonNull
    <T> List<T> get(@NonNull Class<T> entity);

    @NonNull
    <T> Optional<T> get(@NonNull Class<T> entity, long id);

    <T> void put(@NonNull T entity);

    <T> void put(@NonNull List<T> entity);

    void beginTransaction();

    void setTransactionSuccessful();

    void endTransaction();

    <T> int delete(@NonNull Class<T> entity);

    <T> int delete(Class<T> entity, Long... ids);

    <T> boolean delete(@NonNull T entity);

    Observable<Long> changes(@NonNull Class<?> entityClass);
}

File: app/src/main/kotlin/ru/ktoddler/model/repository/LocalRepository.kt
package ru.ktoddler.model.repository

import com.annimon.stream.Optional
import io.reactivex.Observable

interface LocalRepository {

    operator fun <T : Any> get(entity: Class<T>): List<T>

    operator fun <T : Any> get(entity: Class<T>, id: Long): Optional<T>

    fun <T : Any> put(entity: T)

    fun <T : Any> put(entity: List<T>)

    fun beginTransaction()

    fun setTransactionSuccessful()

    fun endTransaction()

    fun <T : Any> delete(entity: Class<T>): Int

    fun <T : Any> delete(entity: Class<T>, vararg ids: Long): Int

    fun <T : Any> delete(entity: T): Boolean

    fun changes(entityClass: Class<*>): Observable<Long>
}
